### PR TITLE
feat(geo-layers):  ImageryLayer for WMS (includes RFC and initial docs)

### PR DESCRIPTION
**Pull request by ibgreen**
For #

#### Background
- WMS support is a notable missing GIS-type feature in deck.gl.
- loaders.gl has now added a wms module with WMS response loaders, so the stage is set.
- [OpenLayers Example](https://openlayers.org/en/latest/examples/wms-image.html)

#### Change List
- Initial RFC.

### CODE REVIEW

File: dev-docs/RFCs/v8.0/image-layer-wms-rfc.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# RFC: ImageLayer and WMS support
+
+- **Authors**: Ib Green
+- **Date**: Jan 12, 2023
+- **Status**: Draft
+
+## Overview
+
+Some services can generate on-demand images that covers an arbitrary specified viewport (rather than set of tiles). WMS services are one of the most common examples of this type of data source and the lack of WMS support in deck.gl is a notable omission.
```
Comment (felixpalmer) [inline dev-docs/RFCs/v8.0/image-layer-wms-rfc.md:9]: How would this work with map tilt & rotation?

File: dev-docs/RFCs/v8.0/image-layer-wms-rfc.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+The ImageLayer should support multiview. A question is whether there should be some reuse between views. The simplest solution is of course that each view issues and independent fetch.
+
+At minimum, the current images for all views must be cached, so that views can be quickly redrawn.
+
+### Parameter pass-through
+
+WMS services typically offer a number of layers, and this is specified through one of many parameters, so there must be a way to provide parameters to the underlying (WMS) URL that can be freely mixed with layer generated paramaters.
+
+
+## Open Issues: Naming
```
[the remark this answers is not in the document]
Comment (kylebarron) [reply dev-docs/RFCs/v8.0/image-layer-wms-rfc.md:59]: I agree that makes more sense than a layer very specific to WMS

File: dev-docs/RFCs/v8.0/image-layer-wms-rfc.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+}
+```
+
+Such a set of data sources would ideally be completely independent of deck.gl,
+and probably best placed in loaders.gl.
+
+
+## Appendix: Potential Future Improvements
+
+### Mosaicing and Client-Side Tiling ideas
```
Comment (kylebarron) [inline dev-docs/RFCs/v8.0/image-layer-wms-rfc.md:160]: Client-side tiling indeed sounds rather involved, but my intuition is that it would help to get more reliable and less variable performance. Re-fetching the entire viewport feels a bit antiquated in terms of architecture and makes the frontend even more dependent on the performance of the backend. WMS systems can be rather slow to respond I think depending on how much data needs to be loaded to render the requested viewport.

I've never worked with WMS personally but my feeling is that it would seem simpler to request tiles via the TileLayer instead of fetching a new image for the viewport every time the user pans. Whenever I've viewed maps that refetch the entire viewport on every pan, the UX has been low

**Review by kylebarron — APPROVED** [review]

File: dev-docs/RFCs/v8.0/image-layer-wms-rfc.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibgreen) [reply dev-docs/RFCs/v8.0/image-layer-wms-rfc.md]: Yes I did consider `ViewportLayer`, I think it is mentioned in the RFC somewhere. My main hesitation was that I feel that name is quite technical and while it makes a lot of sense for deck.gl fiends, it won't make as much sense for the simplest use cases. 

One could of course still provide a simple ImageryLayer subclass that just defines `renderSubLayers()`. 

A footnote is that one thing we discovered what that just passing template URL strings to the bitmap layer doesn't give a good effect, as bounds change immediately and data doesn't change until the async prop loads, so a lot of jumping. I.e. it didn't seem possible to create a good ViewportLayer without managing the image loads separately. See this [codepen](https://codepen.io/jwasilgeo/pen/XWZRKoj)

````
return new BitmapLayer(props, {
        data: null,
        image: props.data,
        bounds: [west, south, east, north]
```

File: dev-docs/RFCs/v8.0/image-layer-wms-rfc.md (hunk 3 above)
Comment (ibgreen) [reply dev-docs/RFCs/v8.0/image-layer-wms-rfc.md]: > I've never worked with WMS personally but my feeling is that it would seem simpler to request tiles via the TileLayer instead of fetching a new image for the viewport every time the user pans. Whenever I've viewed maps that refetch the entire viewport on every pan, the UX has been low

I believe I included the link to the standard OpenLayers tiled WMS example that shows how labeling collapses when the image server is unaware that is serving tiles (zoom out and you get multiple labels per state). Also some GPU backends can render server side and visualize using WMS, and doing multiple calls will crater performance and generate exponential cost increases.

That said, it will work well in many cases and we should also make it trivial to connect tile layer to WMS sources.

File: docs/api-reference/geo-layers/imagery-layer.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
+import {ImageryLayerDemo} from 'website-components/doc-demos/geo-layers';
+
+<TileLayerDemo />
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/imagery-layer.md:3]: Remove the wrong demo and mark as TODO

File: docs/api-reference/geo-layers/imagery-layer.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+<p class="badges">
+  <img src="https://img.shields.io/badge/lighting-yes-blue.svg?style=flat-square" alt="lighting" />
+</p>
+
+# ImageryLayer
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/imagery-layer.md:9]: Can we mark it as experimental, see [GlobeView](https://deck.gl/docs/api-reference/core/globe-view)

File: examples/website/wms/app.tsx [hunk 6]
```diff
@@ -0,0 +1,160 @@
+// deck.gl, MIT license
+
+import React, {PureComponent} from 'react';
+import {render} from 'react-dom';
+import {StaticMap} from 'react-map-gl';
```
Comment (Pessimistress) [inline examples/website/wms/app.tsx:5]: Use newer versions of react and react-map-gl

File: examples/website/wms/app.tsx [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+export const INITIAL_VIEW_STATE = {
+  latitude: 49.254,
+  longitude: -123.13,
+  zoom: 11,
+  maxZoom: 16,
+  pitch: 45,
+  bearing: 0
+};
+
+export default class App extends PureComponent {
```
Comment (Pessimistress) [inline examples/website/wms/app.tsx:24]: Can we use function components?

If you split the example out to another PR I'm happy to help.

File: examples/website/wms/examples.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  zoom: 9,
+  minZoom: 1,
+  maxZoom: 20,
+  pitch: 0,
+  bearing: 0
+};
+
+export const EXAMPLES = {
+  WMS: {
+    'Terrestris(OpenStreetMap)': {
```
Comment (Pessimistress) [inline examples/website/wms/examples.js:21]: I strongly recommend keeping the examples simple and not include too many branches. The hosted version on the website usually have extra features implemented inside `website/src/examples`.

File: examples/website/wms/vite.config.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { defineConfig } from 'vite';
+import fs from 'fs';
+
+/** Run against local source */
```
Comment (Pessimistress) [inline examples/website/wms/vite.config.ts:4]: Use `examples/vite.config.local.mjs`

File: modules/carto/package.json [hunk 10]
```diff
@@ -32,10 +32,10 @@
     "prepublishOnly": "npm run build-bundle && npm run build-bundle -- --env=dev"
   },
   "dependencies": {
-    "@loaders.gl/gis": "^3.3.0-alpha.10",
-    "@loaders.gl/loader-utils": "^3.3.0-alpha.10",
-    "@loaders.gl/mvt": "^3.3.0-alpha.10",
-    "@loaders.gl/tiles": "^3.3.0-alpha.10",
+    "@loaders.gl/gis": "3.3.0-alpha.14",
```
Comment (Pessimistress) [inline modules/carto/package.json:35]: Please split out loaders upgrade to a separate PR.

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+  return nextRequestId++;
+}
+
+let timeoutId;
+
+/** Runs an action in the future, cancels it if the new action is issued before it executes */
+function debounce(fn: Function, ms = 500): void {
+  clearTimeout(timeoutId);
+  timeoutId = setTimeout(() => fn(), ms);
+}
```
Comment (Pessimistress) [inline modules/geo-layers/src/imagery-layer/imagery-layer.ts:205]: These need to be wrapped in each ImageryLayer or you can't use multiple instances.

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+          }, 0);
+        }
+      }
+    });
+  }
+
+  /** Run a getMetadata on the image service */
+  async _loadMetadata(): Promise<void> {
+    this.props.onMetadataLoadStart();
+    const state = this.state as ImageryLayerState;
```
Comment (Pessimistress) [inline modules/geo-layers/src/imagery-layer/imagery-layer.ts:133]: You can type `this.state`, see https://github.com/visgl/deck.gl/blob/8.8-release/modules/layers/src/icon-layer/icon-layer.ts#L141

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+  
+  /** Load an image */
+  async loadImage(reason: string): Promise<void> {
+    const viewports = this.context.deck?.getViewports() || [];
+    if (viewports.length <= 0) {
+      return;
+    }
+
+    // TODO - need to handle multiple viewports
+    const viewport = viewports[0];
```
Comment (Pessimistress) [inline modules/geo-layers/src/imagery-layer/imagery-layer.ts:152]: This may not be the correct viewport. Pass `this.context.viewport` from `updateState` when you call this method.

We have a handful of layers that cannot render in multiple views simultaneously, but at the minimum, a layer should render correctly if it is bound to a single view by `layerFilter`. See https://deck.gl/docs/developer-guide/views#rendering-layers-in-multiple-views

File: docs/api-reference/geo-layers/imagery-layer.md (hunk 4 above)
Comment (ibgreen) [reply docs/api-reference/geo-layers/imagery-layer.md]: Removed the example, not sure how to mark as TODO in markdown

File: docs/api-reference/geo-layers/imagery-layer.md (hunk 5 above)
Comment (ibgreen) [reply docs/api-reference/geo-layers/imagery-layer.md]: Marked as experimental

File: examples/website/wms/app.tsx (hunk 7 above)
Comment (ibgreen) [reply examples/website/wms/app.tsx]: Example broken out to separate PR: https://github.com/visgl/deck.gl/pull/7659

This example is from loaders.gl where being able to test different services is a goal. I agree that a deck.gl example could be much simpler.

File: modules/carto/package.json (hunk 10 above)
Comment (ibgreen) [reply modules/carto/package.json]: loaders upgrade done and landed

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts (hunk 11 above)
Comment (ibgreen) [reply modules/geo-layers/src/imagery-layer/imagery-layer.ts]: debounce etc wrapped

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts (hunk 12 above)
Comment (ibgreen) [reply modules/geo-layers/src/imagery-layer/imagery-layer.ts]: this.state typed

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts (hunk 13 above)
Comment (ibgreen) [reply modules/geo-layers/src/imagery-layer/imagery-layer.ts]: Using `context.viewport`

File: docs/api-reference/geo-layers/imagery-layer.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+##### `onMetadataLoadComplete` (Function, optional)
+
+`onMetadataLoadComplete` called when a tile successfully loads.
+
+- Default: `() => {}`
+
+Receives arguments:
+
+- `tile` (Object) - the [tile](#tile) that has been loaded.
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/imagery-layer.md:138]: No section for this `#tile`?

File: docs/api-reference/geo-layers/imagery-layer.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+- Default: ``
+
+A base URL to a well-known service type, or a full URL template from which the map images should be loaded.
+
+If `props.serviceType` is `'template'`: a URL template. Substrings `{east}` `{north}` `{east}` `{west}` `{south}`, `{width}` and `{height}` if present, will be replaced with a viewports actual bounds and size. `{layers}` will be replaced with content of props.layers.
+
+##### `serviceType` (string, optional)
+
+- Default: `wms`
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/imagery-layer.md:110]:
```suggestion
- Default: `'wms'`
```

File: docs/api-reference/geo-layers/imagery-layer.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+
+## Properties
+
+Inherits all properties from [base `Layer`](/docs/api-reference/core/layer.md).
+
+### Data Options
+
+##### `data` (string, optional)
+
+- Default: ``
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/imagery-layer.md:102]: Remove if there is no default value? Is it really optional?

File: modules/geo-layers/package.json [hunk 17]
```diff
@@ -51,7 +52,7 @@
     "@deck.gl/extensions": "^8.0.0",
     "@deck.gl/layers": "^8.0.0",
     "@deck.gl/mesh-layers": "^8.0.0",
-    "@loaders.gl/core": "^3.0.0",
+    "@loaders.gl/core": "^3.3.0",
```
Comment (Pessimistress) [inline modules/geo-layers/package.json:55]:
```suggestion
    "@loaders.gl/core": "^3.3.1",
```

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+
+    try {
+      this.props.onImageLoadStart(requestId);
+      const image = await this.state.imageSource.getImage({
+        width,
+        height,
+        bbox: bounds,
+        layers: this.props.layers
+      });
+      Object.assign(this.state, {image, bounds, width, height});
```
Comment (Pessimistress) [inline modules/geo-layers/src/imagery-layer/imagery-layer.ts:184]: Call `this.setState`. The method will also trigger `setNeedsRedraw`

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+  serviceType: 'auto',
+  layers: undefined!,
+  onMetadataLoadStart: () => {},
+  onMetadataLoadComplete: () => {},
+  // eslint-disable-next-line
+  onMetadataLoadError: (error: Error) => console.error(error),
+  onImageLoadStart: () => {},
+  onImageLoadComplete: () => {},
+  // eslint-disable-next-line
+  onImageLoadError: (requestId: unknown, error: Error) => console.error(error, requestId)
```
Comment (Pessimistress) [inline modules/geo-layers/src/imagery-layer/imagery-layer.ts:42]: Define [prop types](https://deck.gl/docs/developer-guide/custom-layers/prop-types#built-in-types) to prevent unnecessary updates in React

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts (hunk 18 above)
Comment (Pessimistress) [reply modules/geo-layers/src/imagery-layer/imagery-layer.ts:184]: I think you'll need to check that `requestId` is larger than the last loaded image before overwriting the state.

File: docs/api-reference/geo-layers/imagery-layer.md [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+<script src="https://unpkg.com/@deck.gl/geo-layers@^8.0.0/dist.min.js"></script>
+```
+
+```typescript
+new deck.ImageryLayer({});
+```
+
+## Image Sources
+
+The `ImageryLayer` knows how to URLs for some well-known geospatial services such as WMS, however it is also possible to specify a custom URL template.
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/imagery-layer.md:67]: It's not clear what the purpose of this section is. Are you describing the features that this layer supports, or the features that a valid image source must implement?

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+    } catch (error) {
+      this.context.onError?.(error as Error, this);
+      this.getCurrentLayer()?.props.onImageLoadError(requestId, error as Error);
+    }
+  }
+
+  // HELPERS
+
+  private _nextRequestId: number = 0;
+  private _timeoutId: any;
```
Comment (Pessimistress) [inline modules/geo-layers/src/imagery-layer/imagery-layer.ts:196]: These need to be in `this.state` to transfer to the next instance

File: modules/geo-layers/src/imagery-layer/imagery-layer.ts [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this._loadMetadata();
+        this.debounce(() => this.loadImage('image source changed'), 0);
+      }
+
+      // Some sublayer props may have changed
+    }
+
+    if (changeFlags.viewportChanged) {
+      this.debounce(() => this.loadImage('viewport changed'));
```
Comment (Pessimistress) [inline modules/geo-layers/src/imagery-layer/imagery-layer.ts:90]: There is a caveat when you use debounce.

`context` is shared among all layers and only correct during the call of lifecycle methods. So when a denounced function is actually executed, `context.viewport` may not be correct.

You should grab the value and pass it in:

```js
updateState(params) {
  const {viewport} = this.context;
  // ...
  this.debounce(() => this.loadImage(viewport));
}

**Review by Pessimistress — APPROVED** [review]
